File: src/solutions/2023/08/java/Main.java
package kival.aoc.year2023.day08;

import kival.aoc.core.*;

import java.util.stream.*;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Collections;

public class Main {
   public final boolean HAS_IO = false;
public final boolean HAS_ALTERNATE = false;

   private long gcd(long a, long b) {
      return b == 0 ? a : gcd(b, a % b);
   }

   private long lcm(long a, long b) {
      return (a * b) / gcd(a, b);
   }

   public String part1(String input, boolean isTest) {
      String[] lines = input.split("\n");
      char[] instructions = lines[0].toCharArray();
      Map<String, String[]> maps = new HashMap<>();
      for (int i = 2; i < lines.length; i++) {
         String[] chunks = lines[i].split(" = ");
         String dest = chunks[0];
         String[] lr = chunks[1].substring(1, chunks[1].length() - 1).split(", ");
         maps.put(dest, lr);
      }
      int idx = 0;
      String nav = "AAA";
      while (true) {
         String[] map = maps.get(nav);
         var c = instructions[idx % instructions.length];
         var m = c == 'L' ? 0 : 1;
         nav = map[m];
         idx++;
         if (nav.equals("ZZZ")) break;
      }
      return String.valueOf(idx);
   }

   public String part2(String input, boolean isTest) {
      String[] lines = input.split("\n");
      char[] instructions = lines[0].toCharArray();
      Map<String, String[]> maps = new HashMap<>();
      List<String> navs = new ArrayList<>();
      for (int i = 2; i < lines.length; i++) {
         String[] chunks = lines[i].split(" = ");
         String dest = chunks[0];
         String[] lr = chunks[1].substring(1, chunks[1].length() - 1).split(", ");
         maps.put(dest, lr);
         if (dest.charAt(2) == 'A') {
            navs.add(dest);
         }
      }

      long res = 1;
      for (int i = 0; i < navs.size(); i++) {
         int j = 0;
         while (true) {
            String[] map = maps.get(navs.get(i));
            var c = instructions[j % instructions.length];
            var m = c == 'L' ? 0 : 1;
            navs.set(i, map[m]);
            j++;
            if (navs.get(i).charAt(2) == 'Z') break;
         }
         res = lcm(res, j);
      }
      return String.valueOf(res);
   }

   public void main(String[] args) {
      Run.execute(args, this::part1, this::part2, HAS_ALTERNATE, HAS_IO);
   }
}

File: src/solutions/2023/08/csharp/Main.cs

using Core;

namespace Year2023
{
   public static class Day08
   {
      static SolutionOptions OPTIONS = new()
      {
         HasIO = false,
         HasAlternate = true,
      };

      static private long gcd(long a, long b)
      {
         return b == 0 ? a : gcd(b, a % b);
      }

      static private long lcm(long a, long b)
      {
         return (a * b) / gcd(a, b);
      }

      static string Part1(string input, bool isTest = false)
      {
         var lines = input.Split("\n");
         var instructions = lines[0].Select(x => x == 'L' ? 0 : 1).ToArray();
         var maps = new Dictionary<string, string[]>();
         for (int i = 2; i < lines.Length; i++)
         {
            var chunks = lines[i].Split(" = ");
            var dest = chunks[0];
            var lr = chunks[1][1..(chunks[1].Length - 1)].Split(", ");
            maps[dest] = lr;
         }

         int idx = 0;
         var nav = "AAA";
         while (true)
         {
            var map = maps[nav];
            nav = map[instructions[idx % instructions.Length]];
            idx++;
            if (nav == "ZZZ") break;
         }
         return idx.ToString();
      }

      static string Part2(string input, bool isTest = false)
      {
         var lines = input.Split("\n");
         var instructions = lines[0].Select(x => x == 'L' ? 0 : 1).ToArray();
         var maps = new Dictionary<string, string[]>();
         var navs = new List<string>();
         for (int i = 2; i < lines.Length; i++)
         {
            var chunks = lines[i].Split(" = ");
            var dest = chunks[0];
            var lr = chunks[1][1..(chunks[1].Length - 1)].Split(", ");
            maps[dest] = lr;
            if (dest[2] == 'A')
            {
               navs.Add(dest);
            }
         }

         var res = 1L;
         for (int i = 0; i < navs.Count; i++)
         {
            var j = 0;
            while (true)
            {
               var m = maps[navs[i]];
               navs[i] = m[instructions[j % instructions.Length]];
               j++;
               if (navs[i][2] == 'Z') break;
            }
            res = lcm(res, j);
         }
         return res.ToString();
      }

      static void Main(string[] args)
      {
         Runner.Run(args, Part1, Part2, OPTIONS);
      }
   }
}
